### PR TITLE
Warn about NANs in log_pdf and loss

### LINKED ISSUE #571: How to find cause of nan error in ELBO function
I am trying to modify the `vae.py` example, and the ELBO loss function is returning `nan` before I even do any training steps. My `model` and `guide` functions give sensible output, so I have no way to guess where the error is.  Thanks for your help.

I wouldn't typically open a github issue for this kind of question, but I am not sure what the best place to get support for `pyro` is. Thanks!

### PR DESCRIPTION
**Pull request by fritzo**
Fixes #571 

This adds warnings for:
- NAN and +INF `Trace.*log_pdf()` methods (at each site)
- NAN `.loss()` and `.loss_and_grads()` of in ELBO implementations

Note that this does not print whether errors occurred in the model vs guide, but that information can be seen from the stack trace if a user promotes warnings to errors via
```py
import warnings
warnings.simplefilter("error")
```
or by running a script via `python -W error my_script.py`.

## Tested

I have not added any death tests.

### CODE REVIEW

**Comment by fritzo** [comment]
@nbren12 Do you think this would have helped catch your error faster?

File: pyro/poutine/trace.py [hunk 1]
```diff
@@ -1,8 +1,19 @@
...
 
 import collections
+import warnings
 
 import networkx
+import numpy as np
+
+
+def _warn_if_nan(name, variable):
+    value = variable.data[0]
```
Comment (neerajprad) [inline pyro/poutine/trace.py]: This seems preferable so as to avoid large tensor checks. I suppose this will leave out some small cases where the NaNs/Infs are in different locations and don't propagate across the tensor?

File: pyro/poutine/trace.py [hunk 2]
```diff
@@ -1,8 +1,19 @@
 from __future__ import absolute_import, division, print_function
 
 import collections
+import warnings
 
 import networkx
+import numpy as np
+
+
+def _warn_if_nan(name, variable):
```
Comment (nbren12) [inline pyro/poutine/trace.py]: Is this `name` argument usually something that is easy to interpret?

Comment (fritzo) [reply pyro/poutine/trace.py]: It is identical to the "site_name" in `pyro.sample("site_name", dist.normal, sigma, mu)` [sic] or `pyro.observe("site_name", dist.normal, mu, sigma, x)` [sic].

**Comment by nbren12** [comment]
Thanks for putting this together so fast! I think this should be pretty nice for me. I think it would be really nice to add a short "troubleshooting" or "debugging" page to the docs with a description of how to track down NaN errors in the loss by elevating warning, looking at stack traces, and maybe doing some step by step debugging.

File: pyro/poutine/trace.py (hunk 1 above)
Comment (fritzo) [reply pyro/poutine/trace.py]: We're only checking on `log_pdf` which has a single element. The reduction is already done by `batch_log_pdf()` via `.sum()`, so we do indeed check for NANs everywhere.

File: pyro/poutine/trace.py (hunk 2 above)
Comment (nbren12) [reply pyro/poutine/trace.py]: ok. so it wouldn't tell you whether the problem is in the guide or the model, but that is handled by examining the stack traces as you describe above I guess?

Comment (fritzo) [reply pyro/poutine/trace.py]: Yeah this won't distinguish guide-vs-model unless you examine the stack trace. It would take a lot more plumbing to add that to the warning message. But I do agree that it would help. We'll keep our eye out for better ways to add model-vs-guide info to warning messages.

**Comment by fritzo** [comment]
Ready to merge.
